### PR TITLE
Create CheckBox component for design system

### LINKED ISSUE #2702: (SP: 3) Create CheckBox component aligned with design system
#### Description:
Implement a reusable `CheckBox` component that meets the design system standards. This component will be used throughout the application to maintain consistency in checkbox styles, behavior, and accessibility.

[Figma](https://www.figma.com/design/liJZDYhUnDP15Pi9bipDcv/Space2Study-(Students)?node-id=20967-98389&t=l5PI8JRXpNFKQBr0-4)

#### Requirements:
- **Styles:**
  - Follow design system variables for colors, typography, padding, and other styles.
  - Support different variants (check, middle).
  - Ensure hover, focus, active, and disabled states are defined according to the design system.
  - Provide a label for checkbox 
  
- **Props:**
  - `variant`: Defines type (check, middle).
  - `label`: Provide a label for the checkbox.
  - `labelPosition`: Supports different positions (bottom, top, end)
  -  `color`: Supports different colors (primary, secondary, error, success)
  - `size`: Supports different sizes (sm, md, lg).
  - `loading`: Disables the checkbox when true. When loading, the  checkbox  shouldn't shrink or change its size compared to its default state
  - `disabled`: Disables the checkbox when true.

Don't forget to cover component with tests and add to the storybook

### PR DESCRIPTION
**Pull request by nebby2105**
#2702 
Created Checkbox component with the following configurable props:

1.  color (primary, secondary, success, error)
2.  variant (check, middle)
3.  label
4.  labelPosition (top, bottom, end)
5.  size (sm, md, lg)
6.  loading
7.  disabled

Default Parameters:

Color: primary
Variant: check
Size: md
Label Position: end
Loading: false
Disabled: false

Default parameters:
Screenshot 2024-11-21 at 18 17 49

Color="success"
Screenshot 2024-11-21 at 18 16 50

Color="error"
Screenshot 2024-11-21 at 18 17 04

labelPosition="top"
Screenshot 2024-11-21 at 18 17 19

labelPosition="bottom"
Screenshot 2024-11-21 at 18 17 32

- [x] create CheckBox component 
- [x] add it to storybook
- [x]  cover with tests

### CODE REVIEW

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    disabled || loading ? 's2s-checkbox--disabled' : '',
+    color === 'error' ? 's2s-checkbox--error' : '',
+    color === 'success' ? 's2s-checkbox--success' : '',
+    `s2s-checkbox--${variant}`,
+    `s2s-checkbox--${size}`
+  ]
+    .filter(Boolean)
+    .join(' ')
+
+  const loaderSize = size === 'sm' ? 14 : size === 'md' ? 18 : 20
```
Comment (Made1ra) [inline src/design-system/components/checkbox/CheckBox.tsx]: Extract this nested ternary operation `size === 'md' ? 18 : 20` into an independent statement for better code readability.

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import Checkbox, { CheckboxProps } from '@mui/material/Checkbox'
+import { FC, ReactNode, useState } from 'react'
+import Loader from '~/components/loader/Loader'
+
+import './CheckBox.scss'
+
+interface CheckBoxProps extends Omit<CheckboxProps, 'size'> {
```
Comment (amoutens) [inline src/design-system/components/checkbox/CheckBox.tsx:8]: it is better to rename the interface due to the fact that the name coincides with the MUI props

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          checked={checked}
+          className='s2s-checkbox__input'
+          color={color}
+          data-testid='checkbox-input'
+          disabled={disabled || loading}
+          indeterminate={variant === 'middle' && checked}
+          onChange={handleChange}
+        />
+      )}
+      <span className='s2s-checkbox__label'>{label}</span>
```
Comment (amoutens) [inline src/design-system/components/checkbox/CheckBox.tsx:72]: maybe it is better to use FormControlLabel from MUI

**Review by amoutens — CHANGES_REQUESTED** [review]

File: src/design-system/components/checkbox/CheckBox.tsx (hunk 1 above)
Comment (amoutens) [inline src/design-system/components/checkbox/CheckBox.tsx]: Extract this nested ternary operation into an independent statement.

Fix this small sonar issue

**Review by amoutens — CHANGES_REQUESTED** [review]

File: src/design-system/stories/CheckBox.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import type { Meta, StoryObj } from '@storybook/react'
+import CheckBox from '../components/checkbox/CheckBox'
```
Comment (Made1ra) [inline src/design-system/stories/CheckBox.stories.tsx]: Consider using an absolute import for better readability and maintainability.

**Review by Made1ra — CHANGES_REQUESTED** [review]

**Review by amoutens — APPROVED** [review]

**Review by Made1ra — APPROVED** [review]

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  size = 'md',
+  ...props
+}) => {
+  const [checked, setChecked] = useState<boolean>(false)
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setChecked(event.target.checked)
+  }
+
+  const blockModifiers = [
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: Use `cn` function here, please
Look [here](https://github.com/ita-social-projects/SpaceToStudy-Client/blob/develop/src/design-system/components/button/Button.tsx#L67) for example

**Review by ShadowOfTheSpace — CHANGES_REQUESTED** [review]

**Review by ShadowOfTheSpace — APPROVED** [review]

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import Checkbox, { CheckboxProps } from '@mui/material/Checkbox'
+import { FC, ReactNode, useState } from 'react'
+import Loader from '~/components/loader/Loader'
+import { cn } from '~/utils/cn'
+
+import './CheckBox.scss'
+
+interface S2SCheckBoxProps extends Omit<CheckboxProps, 'size'> {
```
Comment (Mav-Ivan) [inline src/design-system/components/checkbox/CheckBox.tsx]: Change S2SCheckBoxProps just to CheckBoxProps

**Review by Mav-Ivan — APPROVED** [review]

**Pull request merged**